### DISCUSSION TITLE
Get current assembly name (as compile-time constant)

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#6041] [Language Ideas] [answered]
Normally getting an assembly name would be a library, but I'm arguing it should be a language feature.

I regularly find myself needing to get the name of the current assembly (often for logging and /or auditing purposes), which currently requires doing something along the lines of this:

```cs
var assemblyName = GetType().Assembly.GetName().Name;
```

or

```cs
var assemblyName = typeof(Program).Assembly.GetName().Name;
```

both of which feel wrong; I'm getting to the assembly by using reflection on some random type that happens to be defined in that assembly. It's made even more difficult if you want to use top-level statements where there might not be *any* type to reference.

Surely the compiler knows the assembly name of the code it's currently compiling? Assuming it has one, of course. Isn't there some way we can access the name of the current assembly, preferably in a constant way? Perhaps there can be a pseudo constant, such as `Assembly.Current.Name` and `Assembly.Current.FullName`, which would allow it to be used in things like attributes:

```cs
[SomeAttribute(Assembly.Current.Name)]
public void Foo()
{}
```

Bonus: It also feels like a `[CallerAssemblyName]` attribute may be appropriate here:

```cs
public class Logger
{
    public Logger([CallerAssemblyName] string callingAssembly = null)
    {
    }
}
```

**Comment by HaloFour** [comment]
> I regularly find myself needing to get the name of the current assembly (often for logging and /or auditing purposes), which currently requires doing something along the lines of this

There is also `Assembly.GetExecutingAssembly` to get the assembly of the current method and `Assembly.GetCallingAssembly` to get the assembly of the calling method.  But bouncing off of the runtime type handle is probably the most efficient and predictable way to do this.

> Surely the compiler knows the assembly name of the code it's currently compiling?

Is that necessarily true?  I don't know if the situation has changed but in .NET Framework it was possible that the assembly would be linked to a separate manifest which could have a completely different name.

> which would allow it to be used in things like attributes

I'm curious as to what attributes would require the current assembly name?  Any code written to scan such attributes would be using reflection and would already know the assembly based on where they reflected over that attribute.

Shrug, seems at worst to be a very minor nuisance.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/4984.

**Reply by Richiban** [reply]
Thanks 333Fred, although I was hoping to avoid creating an Assembly object the proposal is similar enough that it can be considered the same.
